libevmasm/Assembly.cpp

```diff
@@ -853,20 +887,116 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
+	appendBigEndian(_dest, 2, static_cast<size_t>(_value));
+}
+}
+
+std::tuple<bytes, std::vector<size_t>, size_t> Assembly::createEOFHeader(std::set<uint8_t> const& _referencedSubIds) const
+{
+	bytes retBytecode;
+	std::vector<size_t> codeSectionSizeOffsets;
+	size_t dataSectionSizeOffset;
+
+	retBytecode.push_back(0xef);
+	retBytecode.push_back(0x00);
+	retBytecode.push_back(0x01);										// version 1
+
+	retBytecode.push_back(0x01);										// kind=type
+	appendBigEndianUint16(retBytecode, m_codeSections.size() * 4u);		// length of type section
+
+	retBytecode.push_back(0x02);										// kind=code
+	appendBigEndianUint16(retBytecode, m_codeSections.size());			// placeholder for number of code sections
+
+	for (auto const& codeSection: m_codeSections)
+	{
+		(void) codeSection;
+		codeSectionSizeOffsets.emplace_back(retBytecode.size());
+		appendBigEndianUint16(retBytecode, 0u);							// placeholder for length of code
+	}
+
+	if (!_referencedSubIds.empty())
+	{
+		retBytecode.push_back(0x03);
+		appendBigEndianUint16(retBytecode, _referencedSubIds.size());
+
+		for (auto subId: _referencedSubIds)
+			appendBigEndianUint16(retBytecode, m_subs[subId]->assemble().bytecode.size());
+	}
+
+	retBytecode.push_back(0x04);										// kind=data
+	dataSectionSizeOffset = retBytecode.size();
+	appendBigEndianUint16(retBytecode, 0u);								// length of data
+
+	retBytecode.push_back(0x00);										// terminator
+
+	for (auto const& codeSection: m_codeSections)
+	{
+		retBytecode.push_back(codeSection.inputs);
+		retBytecode.push_back(codeSection.outputs);
+		appendBigEndianUint16(retBytecode, 0xFFFFu);	// TODO: Add stack height calculation
+	}
+
+	return {retBytecode, codeSectionSizeOffsets, dataSectionSizeOffset};
+}
+
 LinkerObject const& Assembly::assemble() const
 {
 	assertThrow(!m_invalid, AssemblyException, "Attempted to assemble invalid Assembly object.");
 	// Return the already assembled object, if present.
 	if (!m_assembledObject.bytecode.empty())
 		return m_assembledObject;
+
 	// Otherwise ensure the object is actually clear.
 	assertThrow(m_assembledObject.linkReferences.empty(), AssemblyException, "Unexpected link references.");
 
-	LinkerObject& ret = m_assembledObject;
-
 	bool const eof = m_eofVersion.has_value();
 	solAssert(!eof || m_eofVersion == 1, "Invalid EOF version.");
 
+	if (!eof)
+		return assembleLegacy();
+	else
+		return assembleEOF();
+}
+
+LinkerObject const& Assembly::assembleLegacy() const
+{
+	assertThrow(!m_eofVersion.has_value(), AssemblyException, "Expect legacy Assembly. Not EOF");
+	assertThrow(!m_invalid, AssemblyException, "Attempted to assemble invalid Assembly object.");
+	// Return the already assembled object, if present.
+	if (!m_assembledObject.bytecode.empty())
+		return m_assembledObject;
+	// Otherwise ensure the object is actually clear.
+	assertThrow(m_assembledObject.linkReferences.empty(), AssemblyException, "Unexpected link references.");
```

Comment:
You could make it even simpler. Here's what I meant in https://github.com/ethereum/solidity/pull/15367#discussion_r1763175805:

```suggestion
	solAssert(!m_eofVersion.has_value());
	solAssert(!m_invalid);
	solAssert(m_assembledObject.bytecode.empty());
	solAssert(m_assembledObject.linkReferences.empty());
```

All of these are things that `assemble()` has already validated so we can be more succint here and just assert. Similar in `assembleEOF()`.

I mean, it does not affect correctness so I'd approve even as is, but just wanted to explain what I meant.

Reply:
Fixed.